# Обзор TensorFlow, PyTorch и Keras

В машинном обучении и глубоком обучении библиотеки TensorFlow, PyTorch и Keras — это мощные инструменты, которые предоставляют удобные интерфейсы для создания и обучения нейросетей. Рассмотрим, чем каждая из них уникальна, а также приведём примеры базового кода для использования в Google Colab.

---



### TensorFlow
**TensorFlow** — это библиотека от Google, широко используемая для обучения и разработки моделей глубокого обучения. Она предоставляет гибкость в создании нейросетей на уровне низкоуровневых операций и также поддерживает автоматизацию с помощью высокоуровневого API Keras (встроенного).

#### Пример: Создание простой нейросети с TensorFlow

In [3]:
# Установка библиотеки (если нужно)
!pip install tensorflow >> None

In [4]:
import tensorflow as tf

# Создание простейшей модели нейросети
model = tf.keras.Sequential([
    tf.keras.layers.Dense(128, activation='relu', input_shape=(784,)),  # Входной слой
    tf.keras.layers.Dense(10, activation='softmax')  # Выходной слой для 10 классов
])

# Компиляция модели
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Проверка модели
model.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                      │ (None, 128)                 │         100,480 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

Мы создаем модель `Sequential`, которая представляет собой последовательность слоев.

- `Dense(128, activation='relu', input_shape=(784,))`: Это первый слой сети. Слой `Dense` — это полносвязный слой (fully connected layer), который имеет 128 нейронов и использует функцию активации ReLU (Rectified Linear Unit). `input_shape=(784,)` указывает на входную форму данных (784 признака), например, для изображений 28x28 пикселей, выровненных в одномерный массив.
  
- `Dense(10, activation='softmax')`: Это выходной слой с 10 нейронами, который использует функцию активации `softmax`. Она нужна для задач классификации, так как позволяет получить вероятности принадлежности к каждому из 10 классов.

`model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])`
Здесь модель компилируется с настройками для обучения:

- `optimizer='adam'`: Оптимизатор Adam, который хорошо подходит для глубокого обучения.
- `loss='sparse_categorical_crossentropy'`: Функция потерь для задачи многоклассовой классификации.
- `metrics=['accuracy']`: Метрика для оценки точности модели.


`model.summary()`

Этот метод выводит структуру модели, включая информацию о каждом слое, его размере и количестве обучаемых параметров.

- **Model: "sequential_1"**: Имя модели.
- **Layer (type)**: Название и тип слоя. Здесь два слоя: `dense_2` и `dense_3`.
  - `dense_2` (Dense) — первый полносвязный слой.
  - `dense_3` (Dense) — выходной полносвязный слой.
- **Output Shape**: Форма выходных данных для каждого слоя.
  - `(None, 128)`: Выходной размер первого слоя имеет 128 нейронов, `None` указывает на неопределенное количество примеров в одном батче.
  - `(None, 10)`: Выходной размер второго слоя, состоящего из 10 нейронов (для 10 классов).
- **Param #**: Количество параметров для обучения в каждом слое.
  - Первый слой имеет 100,480 параметров, потому что \(784 \times 128 + 128 = 100,480\) (веса и смещения).
  - Второй слой имеет 1,290 параметров, потому что \(128 \times 10 + 10 = 1,290\).

### PyTorch
**PyTorch** — библиотека от Facebook, известная своей гибкостью и простотой, особенно популярна среди исследователей. Она предоставляет доступ к тензорам (массивам данных), автоматическому дифференцированию и построению моделей. Отличается динамическим вычислительным графом, который упрощает отладку.

#### Пример: Создание простой нейросети с PyTorch

In [6]:
# Установка библиотеки (если нужно)
!pip install torch >> None

In [7]:
import torch
import torch.nn as nn                   #Содержит модули для создания нейросетей
import torch.optim as optim             #Содержит алгоритмы оптимизации для обновления весов модели.

# Определение простой модели
class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.fc1 = nn.Linear(784, 128)  # Входной слой
        self.fc2 = nn.Linear(128, 10)   # Выходной слой для 10 классов

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.softmax(self.fc2(x), dim=1)
        return x

# Инициализация модели, функции потерь и оптимизатора
model = SimpleNN()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

print(model)

SimpleNN(
  (fc1): Linear(in_features=784, out_features=128, bias=True)
  (fc2): Linear(in_features=128, out_features=10, bias=True)
)


Создается класс SimpleNN, который наследуется от nn.Module, базового класса для всех моделей PyTorch. В методе __init__ определены слои модели:

`self.fc1 = nn.Linear(784, 128)`: Полносвязный слой (fully connected), принимающий на вход 784 нейрона и передающий 128 нейронов на следующий слой. Этот слой работает как входной слой.

`self.fc2 = nn.Linear(128, 10)`: Полносвязный слой, который принимает 128 нейронов и выдает 10 выходов. Он используется как выходной слой, и каждый из 10 нейронов представляет вероятность принадлежности к одному из 10 классов.

Метод `forward` определяет, как данные проходят через сеть:

`x = torch.relu(self.fc1(x))`: Активируется первый слой с функцией активации `ReLU` (Rectified Linear Unit), которая обнуляет отрицательные значения.

`x = torch.softmax(self.fc2(x), dim=1)`: Применяется функция softmax на выходе второго слоя. Она преобразует выходы в вероятности по каждому из 10 классов. Аргумент dim=1 указывает, что softmax применяется по строкам, для каждого примера в батче.

`model = SimpleNN()`: Создается экземпляр модели SimpleNN.

`criterion = nn.CrossEntropyLoss()`: Определяется функция потерь CrossEntropyLoss, которая используется для задач классификации.

`optimizer = optim.Adam(model.parameters(), lr=0.001)`: Настраивается оптимизатор Adam для обновления параметров модели с использованием заданного коэффициента обучения lr=0.001.

**Объяснение результата**

`SimpleNN`: Имя класса модели.

`(fc1): Linear(in_features=784, out_features=128, bias=True)`: Первый слой fc1, являющийся полносвязным (Linear). Он принимает 784 входных признака и выдает 128 выходных значений. Параметр bias=True указывает, что добавляется смещение для каждого выхода.

`(fc2): Linear(in_features=128, out_features=10, bias=True)`: Второй слой fc2, который также является полносвязным. Он принимает 128 входов от предыдущего слоя и выдает 10 значений, соответствующих 10 классам. bias=True также указывает на наличие смещения

### Keras
**Keras** — это высокоуровневый интерфейс, который позволяет быстро и просто строить модели нейросетей. Keras ранее был независимой библиотекой, но с версии TensorFlow 2.0 он интегрирован в TensorFlow, что делает его доступным для работы через `tf.keras`.

#### Пример: Создание простой нейросети с Keras (через TensorFlow)

In [8]:
# Установка библиотеки (если нужно)
!pip install tensorflow >> None

In [10]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Создание модели
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),  # Входной слой
    Dense(10, activation='softmax')  # Выходной слой для 10 классов
])

# Компиляция модели
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Проверка модели
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                      │ (None, 128)                 │         100,480 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

`from tensorflow.keras.models import Sequential`

`from tensorflow.keras.layers import Dense`

Импортируются классы `Sequential` и `Dense` из Keras:

`Sequential`: Используется для создания последовательной модели, где каждый слой добавляется один за другим.

`Dense`: Полносвязный слой, в котором каждый нейрон соединен с каждым нейроном следующего слоя.

Cоздается последовательная модель с двумя слоями:

`Dense(128, activation='relu', input_shape=(784,))`: Первый слой с 128 нейронами и функцией активации ReLU. Параметр input_shape=(784,) определяет, что каждый входной вектор имеет размер 784 (например, это может быть изображение размером 28x28, развернутое в вектор).

`Dense(10, activation='softmax')`: Выходной слой с 10 нейронами и функцией активации softmax, которая используется для многоклассовой классификации, где каждый выходной нейрон представляет вероятность принадлежности к одному из 10 классов.

`model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])`

`optimizer='adam'`: Указывается оптимизатор Adam для обновления весов модели.

`loss='sparse_categorical_crossentropy'`: Задает функцию потерь, которая подходит для задач многоклассовой классификации, где метки представлены целыми числами (например, 0, 1, ..., 9 для 10 классов).

`metrics=['accuracy']`: Указывает, что в процессе обучения будет отслеживаться точность.

**Объяснение результата**

- **Model**: Имя модели (`"sequential_3"`, автоматически сгенерированное Keras).
  
- **Layer (type)**: Имя и тип слоя:
  - `dense_6 (Dense)`: Первый слой `Dense`, содержащий 128 нейронов с активацией ReLU.
  - `dense_7 (Dense)`: Второй слой `Dense`, содержащий 10 нейронов с активацией softmax.

- **Output Shape**: Размер выхода для каждого слоя:
  - `(None, 128)`: Первый слой выдает тензор размера `(None, 128)`. `None` означает, что размер пакета не фиксирован и будет установлен во время обучения или предсказания.
  - `(None, 10)`: Второй слой выдает тензор размера `(None, 10)`, где каждый из 10 выходов представляет вероятность принадлежности к одному из классов.

- **Param #**: Количество параметров (весов и смещений) в каждом слое:
  - `100,480` для первого слоя. Параметры вычисляются как `784 * 128 + 128 = 100,480` (входы * нейроны + смещения).
  - `1,290` для второго слоя. Параметры: `128 * 10 + 10 = 1,290` (выходы предыдущего слоя * нейроны + смещения).

- **Total params**: Общее количество параметров — 101,770, из которых все являются **trainable** (тренируемыми) параметрами, то есть их значения будут обновляться во время обучения.



### Сравнение основных функций

| Функция                        | TensorFlow                        | PyTorch                            | Keras                                   |
|--------------------------------|-----------------------------------|------------------------------------|-----------------------------------------|
| Язык написания                 | C++, CUDA, Python                | Lua                                | Python                                  |
| Архитектура                    | Неудобная для использования      | Сложная, менее читаемая            | Простая, лаконичная, читаемая           |
| Уровень API                    | Высокий и низкий                 | Низкий                             | Высокий                                 |
| Датасеты                       | Большие, высокая производительность | Большие, высокая производительность | Меньшие наборы данных                   |
| Отладка                        | Сложно проводить отладку         | Хорошие возможности отладки        | Простая сеть, отладка редко требуется   |
| Есть ли готовые модели?        | Да                               | Да                                 | Да                                      |
| Популярность                   | Вторая по популярности           | Третья по популярности             | Самая популярная                        |
| Скорость                       | Быстрая, высокая производительность | Быстрая, высокая производительность | Медленная, низкая производительность    |


TensorFlow с Keras удобен для быстрых прототипов, PyTorch — для исследований и сложных кастомизированных моделей, а чистый Keras может быть лучшим для начала изучения глубокого обучения.